int/implicit12.f90

```diff
@@ -0,0 +1,1662 @@
+         Qmax=1.2D0
+      ELSE
+         Qmax=RCNTRL(11)
+      END IF
+
+!~~~>  Check if tolerances are reasonable
+      IF (ITOL == 0) THEN
+         IF (AbsTol(1) <= ZERO .OR. RelTol(1) <= 10.D0*Roundoff) THEN
+            PRINT * , ' Scalar AbsTol = ',AbsTol(1)
+            PRINT * , ' Scalar RelTol = ',RelTol(1)
+            CALL SDIRK_ErrorMsg(-5,Tinitial,ZERO,Ierr)
+         END IF
+      ELSE
+         DO i=1,N
+            IF (AbsTol(i) <= 0.D0.OR.RelTol(i) <= 10.D0*Roundoff) THEN
+              PRINT * , ' AbsTol(',i,') = ',AbsTol(i)
+              PRINT * , ' RelTol(',i,') = ',RelTol(i)
+              CALL SDIRK_ErrorMsg(-5,Tinitial,ZERO,Ierr)
+            END IF
+         END DO
+      END IF
+    
+!~~~>  Special treatment of backward Euler
+    !IF (sdMethod == BEL) THEN
+           StartNewton = .FALSE.
+           FacMin = 1.0_dp
+           FacMax = 1.0_dp
+           FacSafe= 1.0_dp
+           FacRej = 1.0_dp
+           Qmin   = 1.0_dp
+           Qmax   = 1.0_dp
+           Hmax   = Hstart
+           IF ( Hstart <= 100_dp*Roundoff ) THEN
+                   PRINT * , ' Backward Euler calling error:'
+                   PRINT * , '   it requires Hstart > 100*Roundoff'
+                   STOP
+           END IF
+    !END IF
+
+
+    IF (Ierr < 0) RETURN
+
+    CALL SDIRK_Integrator( N,Tinitial,Tfinal,Y,Ierr )
+
+
+   !END SUBROUTINE SDIRK
+
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   CONTAINS  !  PROCEDURES INTERNAL TO SDIRK
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   SUBROUTINE SDIRK_Integrator( N,Tinitial,Tfinal,Y,Ierr )
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+      USE KPP_ROOT_Parameters
+      IMPLICIT NONE
+
+!~~~> Arguments:      
+      INTEGER, INTENT(IN) :: N
+      KPP_REAL, INTENT(INOUT) :: Y(NVAR)
+      KPP_REAL, INTENT(IN) :: Tinitial, Tfinal
+      INTEGER, INTENT(OUT) :: Ierr
+      
+!~~~> Local variables:      
+      KPP_REAL :: Z(NVAR), NewtonRate, SCAL(NVAR), RHS(NVAR),T,Tdirection, H, &
+                  Theta, NewtonPredictedErr,NewtonIncrement, NewtonIncrementOld
+      INTEGER :: j, IER, NewtonIter, IP(NVAR)
+      LOGICAL :: NewtonDone
+      
+#ifdef FULL_ALGEBRA      
+      KPP_REAL FJAC(NVAR,NVAR),  E(NVAR,NVAR)
+#else      
+      KPP_REAL FJAC(LU_NONZERO), E(LU_NONZERO)
+#endif      
+      KPP_REAL, PARAMETER :: ZERO = 0.0d0, ONE = 1.0d0
+
+
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+!~~~>  Initializations
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+      T = Tinitial
+      Tdirection = SIGN(ONE,Tfinal-Tinitial)
+      H = (Tfinal - Tinitial)*Tdirection
+      !print *,'H ',H
+
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+!~~~>  Simplified Newton iterations
+!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+!~~~>  Starting values for Newton iterations
+       CALL Set2zero(N,Z)
+       CALL WADD(N,Y,Z,Z)              ! Zi <- Y + Zi
```

Comment:
@maqsoodrajput @RolfSander I have been running into segmentation faults in the WADD function.  Wonder if this is really needed as we can now add a vector with a direct Z = X + Y statement.  Let me know your thoughts.

Reply:
I was able to fix this with:
```console
   !~~~> In order to remove the prior EQUIVALENCE statements (which
   !~~~> are not thread-safe), we now have declared VAR and FIX as
   !~~~> threadprivate pointer variables that can point to C.
   VAR => C(1:NVAR )
   FIX => C(NVAR+1:NSPEC)

   CALL SDIRK( NVAR,   TIN,   TOUT,   VAR,     RTOL, ATOL,          &
               RCNTRL, ICNTRL,RSTATUS,ISTATUS, Ierr                )

   !~~~> Free pointers
   VAR => NULL()
   FIX => NULL()
```
as VAR and FIX are now local variables in the integrator.